### PR TITLE
Happy new year!

### PR DESCRIPTION
**Pull request by nazmariam**
### Object oriented javascript

[Demo](https://nazmariam.github.io/frogger/index.html) |
[Code base](https://github.com/nazmariam/nazmariam.github.io/blob/master/frogger/index.html)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/nazmariam/oop-js/js/app.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+};
+Player.prototype.toStart = function () {
+    this.x = start.x;
+    this.y = start.y;
+};
+Player.prototype.handleInput = function(code){
+    switch (code) {
+        case 'left':
+            this.x-=100;
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: This is a magic number. Worth being defined as a constant with a descriptive name. Please, define a const for it and other magic numbers across the code.
You can group them in object literals just like you did with `field` and `start`.

File: submissions/nazmariam/oop-js/js/app.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
+const start = {
+    x: 202,
+    y: 390
+};
+const field = {
+    w: 505,
+    h: 606
+};
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: There are two major conventions as to const names.
Please, follow the `ALL_CAPS_WITH_UNDERLINES` convention for now.
Fields inside the object const can still be regular `camelCase`.

File: submissions/nazmariam/oop-js/js/app.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            if (this.y < 0) {
+                win++;
+                console.log(win);
+                alert("Happy New Year! You win "+win+" times!");
+                this.toStart();
+            }
+            break;
+    }
+};
+function checkCollision(enemy) {
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: This function can effectively be a method of `Enemy`.
You will have to refactor it then.
Neither class should refer to global variables like `player` as classes should be agnostic of the outer world. The only exception is global constants and library methods.
So, when you make this fn an `Enemy` method make sure enemies receive reference to the player explicitly, store it in their props and refer to the player via a prop in `.checkCollision`

File: submissions/nazmariam/oop-js/js/app.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+let allEnemies = [
+    new Enemy(-100, 55, 100),
+    new Enemy(-400, 135, 70),
+    new Enemy(0, 135, 35),
+    new Enemy(0, 220, 50),
+    new Enemy(-300, 220, 70)
+];
+let player = new Player();
+document.addEventListener('keyup', function(e) {
+    var allowedKeys = {
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: Semantically this a complex const.

File: submissions/nazmariam/oop-js/js/app.js (hunk 3 above)
Comment (nazmariam) [reply submissions/nazmariam/oop-js/js/app.js]: Did it. Please, check.

File: submissions/nazmariam/oop-js/js/app.js (hunk 4 above)
Comment (nazmariam) [reply submissions/nazmariam/oop-js/js/app.js]: done

File: submissions/nazmariam/oop-js/js/app.js (hunk 2 above)
Comment (nazmariam) [reply submissions/nazmariam/oop-js/js/app.js]: done

File: submissions/nazmariam/oop-js/js/app.js (hunk 4 above)
Comment (OleksiyRudenko) [reply submissions/nazmariam/oop-js/js/app.js]: ... which means it is better declared as `const` )

File: submissions/nazmariam/oop-js/js/app.js (hunk 1 above)
Comment (OleksiyRudenko) [reply submissions/nazmariam/oop-js/js/app.js]: > This is a magic number. Worth being defined as a constant with a descriptive name. Please, define a const for it **and other magic numbers across the code**.
> You can group them in object literals just like you did with field and start.

File: submissions/nazmariam/oop-js/js/app.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+};
+let win = 0;
+let Enemy = function(x,y,speed) {
+    this.sprite = 'images/enemy-bug.png';
+    this.x = x;
+    this.y = y;
+    this.speed = speed;
+};
+Enemy.prototype.update = function(dt,gamer) {
+    this.x += dt*this.speed;
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: Does your code work?

Comment (nazmariam) [reply submissions/nazmariam/oop-js/js/app.js]: it works, cause I put player as gamer in this function in engine.js

Comment (nazmariam) [reply submissions/nazmariam/oop-js/js/app.js]: Fixed without changing engine.

File: submissions/nazmariam/oop-js/js/app.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    if (
+       this.gamer.x < this.x + STEP.x/2 &&
+        this.gamer.x + STEP.x/2 > this.x &&
+        this.gamer.y < this.y + STEP.y/4 &&
+        this.gamer.y + STEP.y/4 > this.y
+    ) {
+        return true;
+    } else {
+        return false;
+    }
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: The fragment above can be optimized.
Expression in `if` condition returns `true` or `false` itself, so you can just return it.
E.g.
```
return this.gamer.x < this.x + STEP.x / 2 &&
  this.gamer.x + STEP.x/2 > this.x &&
  whateverElseLogicalTests
```

File: submissions/nazmariam/oop-js/js/app.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        this.gamer.x = START.x;
+        this.gamer.y = START.y;
+    }
+};
+Enemy.prototype.render = function() {
+    if(this.x >= FIELD.w)
+        this.x = -STEP.x;
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+Enemy.prototype.checkCollision = function(){
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: By conventions functions that return a boolean start with a special verb, which is normally `is`.
Imagine you're asking a question "Is there a collision?" and a response might be "True, there is". Other modal verbs are also in use (`should`, `can`, `does` etc).

You are not required to fix this. Just keep in mind.
**Please, do not spend time on refactoring this**

File: submissions/nazmariam/oop-js/js/app.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    if (this.checkCollision()){
+        alert("You loose");
+        this.gamer.x = START.x;
+        this.gamer.y = START.y;
+    }
+};
+Enemy.prototype.render = function() {
+    if(this.x >= FIELD.w)
+        this.x = -STEP.x;
```
Comment (OleksiyRudenko) [inline submissions/nazmariam/oop-js/js/app.js]: This method is dedicated to drawing. You have different method to manage location. So location boundaries tests is more appropriate to have there.

Separation of Concerns principle helps to write more consistent code.

Fix this in your own game repo. No need to update the PR.

**Pull request merged**